ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/WallClock.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+module Test.Consensus.BlockchainTime.WallClock (tests) where
+
+import           Data.Time.Clock
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.HUnit
+import           Test.Tasty.QuickCheck
+
+import           Control.Monad.Class.MonadSTM.Strict
+
+import           Ouroboros.Consensus.BlockchainTime
+import           Test.Util.Orphans.Arbitrary ()
+
+tests :: TestTree
+tests = testGroup "WallClock" [
+      testProperty "delayNextSlot" prop_delayNextSlot
+    ]
+
+{-------------------------------------------------------------------------------
+  Test for IO
+-------------------------------------------------------------------------------}
+
+-- | Parameters for testing 'timeUntilNextSlot' in some actual IO code
+data TestDelayIO = TestDelayIO {
+      -- | System start
+      --
+      -- Since we don't actually " start " the system in any way, we specify
+      -- this as an offset _before_ the start of the test.
+      tdioStart'  :: NominalDiffTime
+
+      -- | SlotNo length
+      --
+      -- Since this test is run in IO, we will keep the slot length short.
+    , tdioSlotLen :: SlotLength
+    }
+  deriving (Show)
+
+instance Arbitrary TestDelayIO where
+  arbitrary = do
+      tdioStart'  <- arbitrary
+      tdioSlotLen <- slotLengthFromMillisec <$> choose (1, 1_000)
+      return TestDelayIO{..}
+
+-- | Just as a sanity check, also run the tests in IO
+--
+-- We override the maxinum number of tests since there are slow.
```

Comment:
Add comment: this test _can_ fail if process under heavy load (the slot after the delay might be later than the one we expect).